keps/sig-node/24-apparmor-ga/README.md

```diff
@@ -0,0 +1,682 @@
+enforce AppArmor, and prevents the kubelet from needing to resolve annotations &
+fields.
+
+If both AppArmor annotations _and_ fields are specified, the values MUST match.
+This will be enforced in API validation.
+
+To raise awareness of annotation usage (in case of old automation), a warning
+mechanism will be used to highlight that support will be dropped in v1.24.
+The mechanisms being considered are audit annotations, annotations on the
+object, events, or a warning as described in [KEP
+#1693](/keps/sig-api-machinery/1693-warnings).
+
+#### PodSecurityPolicy Enforcement
+
+The PodSecurityPolicy admission controller must continue to check the PSP object
+for annotations, as well as for fields.
+
+When setting default profiles, PSP only needs to set the field. The API
+machinery will handle setting the annotation as necessary.
+
+When enforcing allowed profiles, the PSP should check BOTH the annotations &
+fields. This is implicitly given because annotations as well as the new field
+will be immutable.
+
+#### Pod Update
+
+The AppArmor fields on a pod are immutable, which also applies to the
+[annotation](https://github.com/kubernetes/kubernetes/blob/b46612a74224b0871a97dae819f5fb3a1763d0b9/pkg/apis/core/validation/validation.go#L177-L182).
+
+Therefore, AppArmor annotation updates will be ignored. This maintains backwards
+API compatibility (no tightening validation), and makes a small stabilizing
+change to behavior (new Kubelets will ignore the update).
+
+When an [Ephemeral Container](20190212-ephemeral-containers.md) is added, it
+will follow the same rules for using or overriding the pod's AppArmor profile.
+Ephemeral container's will never sync with an AppArmor annotation.
+
+#### PodTemplates
+
+PodTemplates (e.g. ReplaceSets, Deployments, StatefulSets, etc.) will be
+ignored. The field/annotation resolution will happen on template instantiation.
+
+To raise awareness of existing controllers using the AppArmor annotations that
+need to be migrated, a warning mechanism will be used to highlight that support
+will be dropped in v1.24.
+
+The mechanisms being considered are audit annotations, annotations on the
+object, events, or a warning as described in [KEP
+#1693](/keps/sig-api-machinery/1693-warnings).
+
+#### Runtime Profiles
+
+The API Server will continue to reject annotations with runtime profiles
+different than `runtime/default`, to maintain the existing behavior.
+
+Violations would lead to the error message:
+
+```
+Invalid value: "runtime/profile-name": must be a valid AppArmor profile
+```
+
+#### Kubelet Backwards compatibility
+
+The changes brought to the Kubelet by this KEP will ensure backwards
+compatibility in a similar way the changes above define it at API Server level.
+Therefore, the AppArmor profiles will be applied following the priority order:
+
+1. Container-specific field.
+2. Container-specific annotation.
+3. Pod-wide field.
+4. Pod-wide annotation.
```

Comment:
copy/paste from seccomp? there is no pod-wide annotation

Reply:
Yes, removed that one.